memory-management/src/dynamic_layout.rs

```diff
@@ -0,0 +1,195 @@
+//! Dynamic memory layout containers
+
+/// An array of equally spaced values
+#[derive(Clone)]
+#[repr(C)]
+pub struct DynamicLayoutArray<'a, T> {
+    _element_type: std::marker::PhantomData<&'a T>,
+    element_count: u32,
+    values_stride: u32,
+    values_offset: u32,
+}
+
+impl<'a, T> DynamicLayoutArray<'a, T> {
+    pub fn new_strided(start_offset: usize, element_count: usize, values_stride: usize) -> Self {
+        Self {
+            _element_type: std::marker::PhantomData::default(),
+            element_count: element_count as u32,
+            values_stride: values_stride as u32,
+            values_offset: start_offset as u32,
+        }
+    }
+
+    pub fn new_consequtive(start_offset: usize, element_count: usize) -> Self {
+        Self::new_strided(start_offset, element_count, std::mem::size_of::<T>())
+    }
+
+    pub fn offset_at_index(&self, element_index: usize) -> usize {
+        (self.values_offset as usize)
+            .saturating_add((self.values_stride as usize).saturating_mul(element_index))
+    }
+
+    pub fn start_offset(&self) -> usize {
+        self.values_offset as usize
+    }
+
+    pub fn end_offset(&self) -> usize {
+        self.offset_at_index(self.element_count as usize)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.element_count == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.element_count as usize
+    }
+
+    pub fn get(&self, element_index: usize) -> Option<&T> {
+        if element_index >= self.element_count as usize {
+            return None;
+        }
+        Some(unsafe {
+            &*((self as *const _ as *const u8).add(self.offset_at_index(element_index)) as *const T)
+        })
+    }
+
+    pub fn get_mut(&mut self, element_index: usize) -> Option<&mut T> {
+        if element_index >= self.element_count as usize {
+            return None;
+        }
+        Some(unsafe {
+            &mut *((self as *mut _ as *mut u8).add(self.offset_at_index(element_index)) as *mut T)
```

Comment:
I think you should cast to `usize` and use integer math here, as that discards pointer provenance and avoids UB. I don't remember how you're planning to use this (example in rustdoc pls :P), but add is UB if the result is not inside the same allocation as self.

Reply:
For now, this is indented to be used with the header being in the same allocation as the body. So I would like to keep the pointer provenance.